active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,713 @@
+  }
+}
+```
+
+If extensions require a configuration, there must be a parameters file (Bicep or JSON) that provide the extension
+configuration. If it is not supplied, there should be appropriate error diagnostics. There will be no support for
+inlining extension configurations from CLI arguments.
+
+### The new design from the ARM template perspective
+
+The main Bicep deployment would compile to the ARM template following this paragraph. Only REP relevant data is shown.
+The "extensionConfigs" deployment property will be similar in style to how deployment parameters and type definitions
+are defined in deployment templates. Configuration object property values can be literal values such as strings,
+booleans, and objects. Values can also be directives, such as to fetch a value from an ARM resource API. A directive
+consists of one or more evaluation steps. The extensions property supplied to the deployment PUT will support language
+expressions that will be evaluated against the template defining the deployment. This allows for use of non-secure
+deployment parameters to fill in portions of the configuration (namespace, resource names). The Deployment engine will
+be responsible for validation of expressions to ensure secret values are not processed in them.
+
+main.json - The root deployment
+
+```json5
+{
+  // root deployment
+  "template": {
+    "resources": {
+      "extResources": {
+        // Nested deployment
+        "type": "Microsoft.Resources/deployments",
+        "apiVersion": "2022-09-01",
+        "name": "extResourcesDeploymentName",
+        "properties": {
+          // Language expressions can be used. Expressions are not allowed to reference secure parameters or provide data for config secrets in ways that put the secret security at risk.
+          "extensionConfigs": {
+            "k8s": {
+              "type": "object",
+              // allowed types: "object", "inherit" (more on this in a further section)
+              "value": {
+                "namespace": {
+                  "type": "string",
+                  "value": "[parameters('namespace')]"
+                },
+                "kubeConfig": {
+                  "type": "armApiCall",
+                  "apiReference": {
+                    "method": "GET",
+                    "resourceId": "[ /subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/... ]",
+                    "apiVersion": "2024-02-01",
+                    "path": "/listClusterAdminCredentials",
+                    "query": "[ ?a=1&b=2 ]",
+                    "valueJsonPath": "[ kubeconfig[0].value ]"
+                  }
+                }
+              }
+            }
+          },
+          "template": {
+            "extensions": {
+              "k8s": {
+                "extension": "Kubernetes",
+                "version": "1.0.0"
+                // "config" is removed, it's expected to be provided by properties of the deployment
```

Comment:
We may need to keep the config type information to enable Bicep type validation for remote modules with extension imports, as the remote module artifact is essentially an ARM template.

Reply:
You're right, I left this detail out in this proposal. I meant the value-based portion. I'll update this to include the metadata needed for validation.

Reply:
Fixed.